### PR TITLE
Feat/155 create new summaryinfrastructurecostscsvfombuilder

### PR DESCRIPTION
**Pull request by ArdtK**
## Issue addressed
Solves #157

## Code of conduct
- [x] I HAVE NOT added sensitive or compromised (test) data to the repository.
- [x] I HAVE NOT added vulnerabilities to the repository.
- [x] I HAVE discussed my solution with (other) members of the KOSWAT team.

## What has been done?
- Added a fom and exporter for the infrastructure cost
- Added `koswat_csv_multi_header_fom.py` as this report has > 1 header line

### Checklist
- [x] Tests are either added or updated.
- [x] Branch is up to date with `master`.
- [x] Updated documentation if needed.

## Additional Notes (optional)
N.A.

### CODE REVIEW

File: koswat/core/io/csv/koswat_csv_writer.py [hunk 1]
```diff
@@ -17,7 +17,12 @@ def write(self, fom_instance: KoswatCsvFomProtocol, to_path: Path) -> None:
         def format_line(line: List[str]) -> str:
             return ";".join(map(str, line))
 
-        _lines = list(map(format_line, [fom_instance.headers] + fom_instance.entries))
+        if isinstance(fom_instance.headers[0], list):
+            _lines = list(map(format_line, fom_instance.headers + fom_instance.entries))
+        else:
+            _lines = list(
+                map(format_line, [fom_instance.headers] + fom_instance.entries)
+            )
```
Comment (Carsopre) [inline koswat/core/io/csv/koswat_csv_writer.py:25]:
```suggestion
        _headers = fom_instance.headers if isinstance(fom_instance.headers[0], list) else [fom_instance.headers] 
        _lines = list(map(format_line, _headers + fom_instance.entries))
```

File: koswat/core/io/csv/koswat_csv_multi_header_fom.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+from koswat.core.io.csv.koswat_csv_fom_protocol import KoswatCsvFomProtocol
+
+
+class KoswatCsvMultiHeaderFom(KoswatCsvFomProtocol):
+    headers: list[list[str]]
+    entries: list[list[str]]
+
+    def __init__(self) -> None:
+        self.headers = []
+        self.entries = []
```
Comment (Carsopre) [inline koswat/core/io/csv/koswat_csv_multi_header_fom.py:10]: If you are not going to use the `__init__` you might as well make this into a `dataclass` with default values (you might have noticed I'm doing it whenever I come upon such initializers).

File: koswat/cost_report/infrastructure/infrastructure_location_profile_cost_report.py [hunk 3]
```diff
@@ -25,15 +24,22 @@ class InfrastructureLocationProfileCostReport(CostReportProtocol):
     infrastructure_location_costs: InfrastructureLocationCosts
 
     @property
-    def location(self) -> Point:
+    def location(self) -> PointSurroundings:
         if not self.infrastructure_location_costs:
-            return None
-        return self.infrastructure_location_costs.location.location
+            return PointSurroundings()
+        return self.infrastructure_location_costs.location
```
Comment (Carsopre) [inline koswat/cost_report/infrastructure/infrastructure_location_profile_cost_report.py:30]: Why is this change required? I do no believe you require such information anymore (you do have the `reinforced_profile` and `infrastructure` define in this class. 
Instead you should just return the coordinates related to this report. If there is not such point then just return `None`.

File: koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+class SummaryInfrastructureCostsCsvFomBuilder(BuilderProtocol):
+    koswat_summary: KoswatSummary
+    _ordered_profile_types: list[str]
+    _ordered_infra_types: list[str]
+
+    def __init__(self) -> None:
+        self.koswat_summary = None
+        self._ordered_profile_types = []
+        self._ordered_infra_types = []
```
Comment (Carsopre) [inline koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py:30]: `dataclass`

File: koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            for _profile_type in self._ordered_profile_types:
+                _totals.append(
+                    sum(
+                        x.total_cost
+                        for x in _infra_cost_per_loc_dict[location][
+                            _profile_type
+                        ].values()
+                    )
+                )
+            return _totals
```
Comment (Carsopre) [inline koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py:75]: so ?

```suggestion
            return list(
                    sum(
                        x.total_cost
                        for x in _infra_cost_per_loc_dict[location][
                            _profile_type
                        ].values()
                    )
                    for _profile_type in self._ordered_profile_types
                )
```


or ...

```suggestion
            def total_cost_per_profile_and_location(profile_type) -> float:
                return sum(map(lambda x: x.total_cost, _infra_cost_per_loc_dict[location][profile_type].values()))

            return list(map(total_cost_per_profile_and_location, self._ordered_profile_types))
```

File: koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        ] = defaultdict(
+            lambda: defaultdict(lambda: defaultdict(InfrastructureLocationCosts))
+        )
+        for _lpr in locations_profile_report_list:
+            for _ilpcr in _lpr.infra_multilocation_profile_cost_report:
+                _infra_location_cost_dict[
+                    _ilpcr.infrastructure_location_costs.location
+                ][_lpr.profile_type_name][
+                    _ilpcr.infrastructure.infrastructure_name
+                ] = _ilpcr.infrastructure_location_costs
```
Comment (Carsopre) [inline koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py:143]: Perhaps it's the typography, but this became unreadable to me.
```suggestion
        def update_infra_location_cost_dict(profile_type_name: str, location: Point, infra_name: str, costs: float):
            _infra_location_cost_dict[location][profile_type_name][infra_name] = costs

        for _lpr in locations_profile_report_list:
            for _ilpcr in _lpr.infra_multilocation_profile_cost_report:
                update_infra_location_cost_dict(
                     _lpr,
                     _ilpcr.infrastructure_location_costs.location,
                     _ilpcr.infrastructure.infrastructure_name,
                     _ilpcr.infrastructure_location_costs
                )
```

**Review by Carsopre — CHANGES_REQUESTED** [review]
Some changes required for this to be merged back to `master`.

File: koswat/core/io/csv/koswat_csv_writer.py (hunk 1 above)
Comment (ArdtK) [reply koswat/core/io/csv/koswat_csv_writer.py:25]: Nice suggestion

File: koswat/cost_report/infrastructure/infrastructure_location_profile_cost_report.py (hunk 3 above)
Comment (ArdtK) [reply koswat/cost_report/infrastructure/infrastructure_location_profile_cost_report.py:30]: We need the `section` in the report, which is not in `self.infrastructure_location_costs.location.location`.
Returning `None` is still fine. Adjusted.

File: koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py (hunk 4 above)
Comment (ArdtK) [reply koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py:30]: Adjusted

File: koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py (hunk 5 above)
Comment (ArdtK) [reply koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py:75]: Thanks, like the first more.

File: koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py (hunk 6 above)
Comment (ArdtK) [reply koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py:143]: Better!

File: koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+        _csv_fom = KoswatCsvMultiHeaderFom()
+        _csv_fom.headers = self._get_headers()
+        _csv_fom.entries = self._get_locations_matrix(
+            self.koswat_summary.reinforcement_per_locations,
+            self.koswat_summary.locations_profile_report_list,
+        )
+
+        return _csv_fom
+
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py:58]: Adjusted

File: koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_exporter.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+            raise ValueError("No 'KoswatSummary' object provided.")
+        if not isinstance(export_path, Path):
+            raise ValueError("No export path location provided.")
+
+        _csv_fom_builder = SummaryInfrastructureCostsCsvFomBuilder()
+        _csv_fom_builder.koswat_summary = koswat_summary
+        _csv_fom = _csv_fom_builder.build()
+        if not _csv_fom:
+            logging.error(
+                "Export of KoswatSummary failed, no FileObjectModel was generated."
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_exporter.py:24]: It was. Adjusted on multiple locations now.

File: tests/cost_report/io/summary/conftest.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                location=_location,
+                available_measures=_a_measures,
+                selected_measure=_selected_measure,
+            )
+        )
+    return _matrix
+
+
+@pytest.fixture
+def valid_mocked_summary() -> KoswatSummary:
```
Comment (Carsopre) [inline tests/cost_report/io/summary/conftest.py:146]:
```suggestion
@pytest.fixture(name="valid_mocked_summary")
def _get_valid_mocked_summary_fixture() -> KoswatSummary:
```

File: tests/cost_report/io/summary/summary_locations/test_summary_locations_csv_exporter.py [hunk 10]
```diff
@@ -8,7 +8,7 @@
 )
 from koswat.cost_report.summary.koswat_summary import KoswatSummary
 from tests import test_results
-from tests.cost_report.io.summary import valid_mocked_summary
+from tests.cost_report.io.summary.conftest import valid_mocked_summary
```
Comment (Carsopre) [inline tests/cost_report/io/summary/summary_locations/test_summary_locations_csv_exporter.py:11]:
```suggestion

```

File: tests/cost_report/io/summary/summary_locations/test_summary_locations_csv_fom_builder.py [hunk 11]
```diff
@@ -4,7 +4,7 @@
     SummaryLocationsCsvFomBuilder,
 )
 from koswat.cost_report.summary.koswat_summary import KoswatSummary
-from tests.cost_report.io.summary import valid_mocked_summary
+from tests.cost_report.io.summary.conftest import valid_mocked_summary
```
Comment (Carsopre) [inline tests/cost_report/io/summary/summary_locations/test_summary_locations_csv_fom_builder.py:7]:
```suggestion

```

File: tests/cost_report/io/summary/test_koswat_summary_exporter.py [hunk 12]
```diff
@@ -6,7 +6,7 @@
 from koswat.cost_report.io.summary.koswat_summary_exporter import KoswatSummaryExporter
 from koswat.cost_report.summary.koswat_summary import KoswatSummary
 from tests import test_results
-from tests.cost_report.io.summary import valid_mocked_summary
+from tests.cost_report.io.summary.conftest import valid_mocked_summary
```
Comment (Carsopre) [inline tests/cost_report/io/summary/test_koswat_summary_exporter.py:9]:
```suggestion

```

File: tests/cost_report/io/summary/summary_infrastructure_costs/test_summary_infrastructure_costs_csv_fom_builder.py [hunk 13]
```diff
@@ -4,20 +4,21 @@
     SummaryInfrastructureCostsCsvFomBuilder,
 )
 from koswat.cost_report.summary.koswat_summary import KoswatSummary
-from tests.cost_report.io.summary import valid_mocked_summary
+from tests.cost_report.io.summary.conftest import valid_mocked_summary
```
Comment (Carsopre) [inline tests/cost_report/io/summary/summary_infrastructure_costs/test_summary_infrastructure_costs_csv_fom_builder.py:7]:
```suggestion

```

File: tests/cost_report/io/summary/summary_infrastructure_costs/test_summary_infrastructure_costs_csv_exporter.py [hunk 14]
```diff
@@ -8,7 +8,7 @@
 )
 from koswat.cost_report.summary.koswat_summary import KoswatSummary
 from tests import test_results
-from tests.cost_report.io.summary import valid_mocked_summary
+from tests.cost_report.io.summary.conftest import valid_mocked_summary
```
Comment (Carsopre) [inline tests/cost_report/io/summary/summary_infrastructure_costs/test_summary_infrastructure_costs_csv_exporter.py:11]:
```suggestion

```

File: tests/cost_report/io/summary/summary_costs/test_summary_costs_fom_builder.py [hunk 15]
```diff
@@ -4,7 +4,7 @@
     SummaryCostsCsvFomBuilder,
 )
 from koswat.cost_report.summary.koswat_summary import KoswatSummary
-from tests.cost_report.io.summary import valid_mocked_summary
+from tests.cost_report.io.summary.conftest import valid_mocked_summary
```
Comment (Carsopre) [inline tests/cost_report/io/summary/summary_costs/test_summary_costs_fom_builder.py:7]:
```suggestion

```

File: tests/cost_report/io/summary/summary_costs/test_summary_costs_csv_exporter.py [hunk 16]
```diff
@@ -8,7 +8,7 @@
 )
 from koswat.cost_report.summary.koswat_summary import KoswatSummary
 from tests import test_results
-from tests.cost_report.io.summary import valid_mocked_summary
+from tests.cost_report.io.summary.conftest import valid_mocked_summary
```
Comment (Carsopre) [inline tests/cost_report/io/summary/summary_costs/test_summary_costs_csv_exporter.py:11]: with `conftest` this is no longer needed.
```suggestion

```

**Review by Carsopre — CHANGES_REQUESTED** [review]
Some last remarks required.

File: tests/cost_report/io/summary/summary_costs/test_summary_costs_csv_exporter.py (hunk 16 above)
Comment (ArdtK) [reply tests/cost_report/io/summary/summary_costs/test_summary_costs_csv_exporter.py:11]: Got it

File: tests/cost_report/io/summary/conftest.py (hunk 9 above)
Comment (ArdtK) [reply tests/cost_report/io/summary/conftest.py:146]: What is the advantage doing it this way? It makes finding the references a bit harder.

Comment (Carsopre) [reply tests/cost_report/io/summary/conftest.py]: just consistency on how we are doing everywhere else, but you might leave it as it is.

**Review by Carsopre — APPROVED** [review]
